src/org/sosy_lab/cpachecker/cfa/mutation/strategy/GenericCFAMutationStrategy.java

```diff
@@ -0,0 +1,368 @@
+    return true;
+  }
+
+  private boolean setLevel(ParseResult pParseResult) {
+    if (depth == -1) {
+      return initLevel(pParseResult);
+    }
+
+    previousMutations.clear();
+    depth++;
+    Collection<ObjectKey> level = getAllObjects(pParseResult);
+    levelSize = level.size();
+    for (ObjectKey o : level) {
+      if (!objectsBefore.contains(o)) {
+        stats.objectsAppearedDuringPass.inc();
+      }
+    }
+
+    if (batchSize <= 1 || levelSize == 0) {
+      countRemained(pParseResult);
+      return false;
+    }
+
+    if (rate < 2) {
+      rate = 2;
+    } else {
+      rate *= 2;
+    }
+
+    if (rate >= levelSize) {
+      batchSize = 1;
+      rate = levelSize;
+    } else {
+      batchSize = (levelSize - 1) / rate + 1;
+    }
+    batchNum = 0;
+    batchCount = rate;
+
+    logger.logf(Level.INFO, "new level: %d deltas of size %d at depth %d", rate, batchSize, depth);
+    return true;
+  }
+
+  private void countRemained(ParseResult pParseResult) {
+    Collection<ObjectKey> objectsAfter = getAllObjects(pParseResult);
+    objectsAfter.removeAll(previousPasses);
+    for (ObjectKey o : objectsAfter) {
+      if (objectsBefore.contains(o)) {
+        logger.logf(Level.INFO, "remained after pass: %s", o);
+        previousPasses.add(o);
+      } else {
+        logger.logf(Level.INFO, "appeared after pass: %s", o);
+        stats.objectsAppearedAfterPass.inc();
+      }
+    }
+    stats.objectsRemainedAfterPass.setNextValue(objectsAfter.size());
+  }
+
+  @Override
+  public void rollback(ParseResult pParseResult) {
+    stats.rollbacks.inc();
+
+    switch (state) {
+      case RemoveComplementNoRollback:
+        state = State.RemoveComplement;
+        previousMutations.clear();
+        int count = 0;
+        for (final ObjectKey object : currentMutation) {
+          if (++count > batchSize) {
+            break;
+          }
+          previousMutations.add(object);
+        }
+        break;
+      case RemoveDeltaNoRollback:
+        state = State.RemoveDelta;
+        break;
+      default:
+        throw new UnsupportedOperationException("" + state);
+    }
+
+    Iterator<RollbackInfo> it = rollbackInfos.iterator();
+    while (it.hasNext()) {
+      final RollbackInfo ri = it.next();
+      //      logger.logf(Level.INFO, "returning ri %s", ri);
+      returnObject(pParseResult, ri);
+    }
+
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public void makeAftermath(ParseResult pParseResult) {
+    Collection<ObjectKey> objects = getAllObjects(pParseResult);
+    objects.removeIf(o -> previousPasses.contains(o));
+    stats.objectsForNextPass.setNextValue(objects.size());
+  }
+
+  @Override
+  public String toString() {
+    return this.getClass().getSimpleName() + "(" + pass + " pass, depth " + depth + ")";
+  }
+
+  @Override
+  public void collectStatistics(Collection<Statistics> pStatsCollection) {
+    pStatsCollection.add(stats);
+    stats =
```

Comment:
Why we reset statistics?

Reply:
To count objects at each pass separately, if strategy is used by Repeated (aka Cycle) Strategy.

Reply:
Looks very overloaded. Do we need all these separate statistics loops?